### PR TITLE
[tune] use default anonymous metric `_metric` if at least a mode is set

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Ray Tune already supports an anonymous metric if e.g. `tune.report(0.5)` is called or a scalar value is `return`ed in the (function) trainable. However, it is currently not used automatically, i.e. users will have to specify `tune.run(train, mode="max", metric="_metric")` to use this metric in the searchers, schedulers, progress reporting, and experiment analysis.

This PR changes this behavior so that the default `_metric` is used when no other metric has been specified but a `mode` has been set. Thus, the API reduces to `tune.run(train, mode="max")`.

I think it's sensible to require setting a mode, as we cannot assume if we're optimizing for higher or lower values. Other libraries make this assumption, or default to one direction. Also, by explicitly setting a mode but _not_ a metric, we can assume that the user actually wants to optimize an anonymous metric.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/schedulers/median_stopping_rule.py [hunk 1]
```diff
@@ -22,7 +23,8 @@ class MedianStoppingRule(FIFOScheduler):
             `training_iteration` as a measure of progress, the only requirement
             is that the attribute should increase monotonically.
         metric (str): The training result objective value attribute. Stopping
-            procedures will use this attribute.
+            procedures will use this attribute. If None but a mode was passed,
+            the anonymous metric `_metric` will be used per default.
```
Comment (richardliaw) [inline python/ray/tune/schedulers/median_stopping_rule.py:27]: (same)

File: python/ray/tune/schedulers/pbt.py [hunk 2]
```diff
@@ -141,7 +141,8 @@ class PopulationBasedTraining(FIFOScheduler):
             `training_iteration` as a measure of progress, the only requirement
             is that the attribute should increase monotonically.
         metric (str): The training result objective value attribute. Stopping
-            procedures will use this attribute.
+            procedures will use this attribute. If None but a mode was passed,
+            the anonymous metric `_metric` will be used per default.
```
Comment (richardliaw) [inline python/ray/tune/schedulers/pbt.py:145]: (same)

File: python/ray/tune/suggest/ax.py [hunk 3]
```diff
@@ -42,10 +43,11 @@ class AxSearch(Searcher):
             (list of two values, lower bound first), "values" for choice
             parameters (list of values), and "value" for fixed parameters
             (single value).
-        objective_name (str): Name of the metric used as objective in this
+        metric (str): Name of the metric used as objective in this
```
Comment (richardliaw) [inline python/ray/tune/suggest/ax.py:46]: (same)

also, maybe do `Optional[str]`

File: python/ray/tune/analysis/experiment_analysis.py [hunk 4]
```diff
@@ -33,7 +33,8 @@ class Analysis:
         experiment_dir (str): Directory of the experiment to load.
         default_metric (str): Default metric for comparing results. Can be
             overwritten with the ``metric`` parameter in the respective
-            functions.
+            functions. If None but a mode was passed, the anonymous metric
+            `_metric` will be used per default.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/tune/analysis/experiment_analysis.py:37]: Optional[str]?

File: python/ray/tune/suggest/hyperopt.py [hunk 5]
```diff
@@ -177,14 +180,22 @@ def __init__(
                     UNRESOLVED_SEARCH_SPACE.format(
                         par="space", cls=type(self)))
                 space = self.convert_search_space(space)
-            self.domain = hpo.Domain(lambda spc: spc, space)
+            self._space = space
+            self.setup_hyperopt()
+
+    def setup_hyperopt(self):
```
Comment (richardliaw) [inline python/ray/tune/suggest/hyperopt.py:186]: maybe do a private method?

**Review by richardliaw — COMMENTED** [review]
looks like a kosher change though tests are failing?

Mainly comments regarding docstring.

File: python/ray/tune/analysis/experiment_analysis.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/analysis/experiment_analysis.py:37]: We're not including the `Optional` keyword in the other docstring references. Let's do this in one go sometime? This doesn't seem to be the right PR to start with it

**Comment by krfricke** [comment]
Failing test is `test_trial_scheduler_pbt` and the error seems unrelated to this PR (occurs in other PRs, too)

File: python/ray/tune/analysis/experiment_analysis.py (hunk 4 above)
Comment (richardliaw) [inline python/ray/tune/analysis/experiment_analysis.py:37]:
```suggestion
            `ray.tune.result.DEFAULT_METRIC` will be used per default.
```

**Review by richardliaw — APPROVED** [review]

**Pull request merged**